packages/ai-jsx/src/react/completion.tsx

```diff
@@ -0,0 +1,97 @@
+import React from 'react';
+import { ChatCompletion, SystemMessage, UserMessage } from '../core/completion.js';
+
+function reactComponentName(component: React.JSXElementConstructor<any> | string) {
+  return typeof component === 'string' ? component : component.name;
+}
+
+export async function UICompletion(
+  { example, children }: { example: React.ReactNode; children: AI.Node },
+  { render, logger }: AI.ComponentContext
+) {
+  const reactComponents = new Set<React.JSXElementConstructor<any> | string>();
+  function collectComponents(node: React.ReactNode | AI.Node, inReact: boolean) {
+    if (Array.isArray(node)) {
+      node.forEach((node) => collectComponents(node, inReact));
+    }
+
+    if (React.isValidElement(node)) {
+      if (inReact) {
+        reactComponents.add(node.type);
+      }
+
+      const childrenAreReact = (inReact || node.type === AI.React) && node.type !== AI.jsx;
+      if ('children' in node.props) {
+        collectComponents(node.props.children, childrenAreReact);
+      }
+    }
+
+    if (AI.isElement(node)) {
+      const childrenAreReact = (inReact || node.tag === AI.React) && node.tag !== AI.jsx;
+      if ('children' in node.props) {
+        collectComponents(node.props.children, childrenAreReact);
+      }
+    }
+  }
+
+  collectComponents(example, true);
+
+  const modelResult = await render(
+    <ChatCompletion>
+      <SystemMessage>
+        You are an AI who is an expert UI designer. The user will provide content in the form of text and you will
+        respond using a set of React components to create a UI for the content. Here are the only available React
+        components and how they should be used:
+        {'\n'}
+        <AI.React>{example}</AI.React>
+        {'\n'}
+        Respond with a JSON object that encodes your UI. The JSON object should match this TypeScript interface:
+        interface Element {'{'}
+        name: string; children: (string | Element)[]
+        {'}'}
+        For example:{'\n'}
+        {'<'}SomeComponent /{'>'}
+        Becomes: {'{'} "name": "SomeComponent", "children": []{'}'}. Respond with only the JSON. Do not include any
+        explanatory prose. Do not use any elements (including HTML elements) other than the ones above.
+      </SystemMessage>
+      <UserMessage>{children}</UserMessage>
+    </ChatCompletion>
+  );
+
+  const validComponents = Object.fromEntries(Array.from(reactComponents).map((c) => [reactComponentName(c), c]));
+
+  interface SerializedComponent {
+    name: string;
+    children: (string | SerializedComponent)[];
+  }
+
+  function toComponent(serializedComponent: SerializedComponent | string) {
+    if (!serializedComponent || typeof serializedComponent === 'string') {
+      return serializedComponent;
+    }
+
+    const Component = validComponents[serializedComponent.name];
+    if (!Component) {
+      logger.warn({ serializedComponent }, `Component not found for ${serializedComponent.name}`);
```

Comment:
This message won't be actionable on its own to the user. Can we include more detail about what the user can do to resolve this?

```suggestion
      logger.warn({ serializedComponent, validComponentNames: Object.keys(validComponents) }, `Component not found for ${serializedComponent.name}`);
```